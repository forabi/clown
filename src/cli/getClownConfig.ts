/* The reason this function is in its own module is so it becomes
easier to mock it with Jest. */
export function getClownConfig(path: string) {
  return require(path);
}
