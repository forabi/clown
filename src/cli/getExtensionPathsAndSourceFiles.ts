import path from 'path';
import { verifyClownConfigLooksGood } from './utils';
import { ClownConfig, ExtensionPathAndSourceFiles } from './types';
import glob from 'globby';
import bluebird from 'bluebird';

export async function getExtensionPathsAndSourceFiles(cwd: string) {
  /* for HEC to work, the user has to have a file called `clown.js` at the location
  where they are running the script */
  const clownConfig = require(path.resolve(cwd, 'clown.js')) as ClownConfig;

  /* We expect our `clownConfig` to look like this:

  {
    extensionsModule: '@hollowverse/config',
    extensions: ['basics', 'typescript', 'eslint']
  }

  Let's verify that or throw */
  verifyClownConfigLooksGood(clownConfig);

  /* The `extensionsModule` key of `clownConfig` tells us where we should find the repo
  that has all the extensions. So, let's get the absolute path to `extensionsModule`. */
  const extensionsModulePath = path.resolve(
    cwd,
    'node_modules',
    clownConfig.extensionsModule,
  );

  /* `extensionsModule` is expected to have a folder named `clown` inside it. Inside `clown` folder,
  there will be one or more folders that correspond to the items in the `extensions` array of the
  `clownConfig`. So let's create a list of the absolute paths to all of those folders. */
  const extensionPaths = clownConfig.extensions.map(extensionName => {
    return path.resolve(extensionsModulePath, extensionName);
  });

  /* Now, `extensionsPaths` is an array of absolute paths to the extension folders. We need a list of all the files and
  folders inside each of the extension folders. We can use `glob` to list the content.

  So, let's map the list of extension paths so that for each path, we also have a list of the files
  paths relative to the extension path:

  [
    [
      "some/path/node_modules/@hollowverse/config/eslint", [".eslintrc.json", "package.json"],
    ],
    [
      "etc...": ["etc...", "etc..."]
    ]
  ]
  */
  return await bluebird.map(extensionPaths, async (extensionPath: string) => {
    return [
      extensionPath,
      await glob('**', { dot: true, cwd: extensionPath }),
    ] as ExtensionPathAndSourceFiles;
  });
}
