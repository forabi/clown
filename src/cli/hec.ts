#!/usr/bin/env node
import path from 'path';
import {
  verifyHecConfigLooksGood,
  isMergeableJsonContent,
  readFile,
  writeFile,
  isDotIgnoreFile,
} from './utils';
import { HecConfig, ExtensionPathsAndSourceFile } from './types';
import glob from 'globby';
import bluebird from 'bluebird';
import _ from 'lodash';
import realmkdirp from 'mkdirp';
import { promisify } from 'util';
import { extendJson } from '../functions/extendJson';
import { extendDotIgnore } from '../functions/extendDotIgnore';

const mkdirp = promisify(realmkdirp);

async function main() {
  /* pwd is the location from where the user is running this shell script */
  const pwd = process.cwd();

  /* for HEC to work, the user has to have a file called `hec.js` at the location
  where they are running the script */
  const hecConfig = require(path.resolve(pwd, 'hec.js')) as HecConfig;

  /* We expect our `hecConfig` to look like this:

  {
    extensionsModule: '@hollowverse/config',
    extensions: ['basics', 'typescript', 'eslint']
  }

  Let's verify that or throw */
  verifyHecConfigLooksGood(hecConfig);

  /* The `extensionsModule` key of `hecConfig` tells us where we should find the repo
  that has all the extensions. So, let's get the absolute path to `extensionsModule`. */
  const extensionsModulePath = path.resolve(
    pwd,
    'node_modules',
    hecConfig.extensionsModule,
  );

  /* `extensionsModule` is expected to have a folder named `hec` inside it. Inside `hec` folder,
  there will be one or more folders that correspond to the items in the `extensions` array of the
  `hecConfig`. So let's create a list of the absolute paths to all of those folders. */
  const extensionPaths = hecConfig.extensions.map(extensionName => {
    return path.resolve(extensionsModulePath, 'hec', extensionName);
  });

  /* Now, `extensionsPaths` is an array of absolute paths to the extension folders. We need a list of all the files and
  folders inside each of the extension folders. We can use `glob` to list the content.

  So, let's map the list of extension paths so that for each path, we also have a list of the files
  paths relative to the extension path:

  [
    [
      "some/path/node_modules/@hollowverse/config/eslint", [".eslintrc.json", "package.json"],
    ],
    [
      "etc...": ["etc...", "etc..."]
    ]
  ]
  */
  const extensionPathsAndSourceFiles = await bluebird.map(
    extensionPaths,
    async (extensionPath: string) => {
      return [
        extensionPath,
        await glob('**', { dot: true, cwd: extensionPath }),
      ] as ExtensionPathsAndSourceFile;
    },
  );

  /* Now we can iterate over the source files for each extension, and either merge them with existing
  destination files, or simply copy them over as-is.

  Since we're doing this operation in an async manner, we return a promise from our `main` function
  to make sure the process doesn't exit before all promises have been resolved.
  (I'm not sure about this last bit, but I'm doing it just in case.) */
  await bluebird.each(extensionPathsAndSourceFiles, iterateOverSourceFiles);

  function iterateOverSourceFiles([
    extensionPath,
    sourceFiles,
  ]: ExtensionPathsAndSourceFile) {
    return bluebird.each(sourceFiles, sourceFile =>
      moveAndExtend(extensionPath, sourceFile),
    );
  }

  async function moveAndExtend(extensionPath: string, sourceFile: string) {
    const sourceFilePath = path.resolve(extensionPath, sourceFile);
    const destinationFilePath = path.resolve(pwd, sourceFile);
    const sourceFileContent = (await readFile(sourceFilePath)) as string;
    const destinationFileContent = await readFile(destinationFilePath);

    if (destinationFileContent === undefined) {
      /* if the destinationFileContent is undefined, that means we couldn't find the destination
      file at the expected location, either because the directory itself or the file doesn't exist.
      So before we attempt to create the file, let's ensure that the directory exists and then create the file. */
      await mkdirp(path.dirname(destinationFilePath));
      return writeFile(destinationFilePath, sourceFileContent);
    }

    if (isDotIgnoreFile(path.basename(destinationFilePath))) {
      const extendedContent = extendDotIgnore(
        destinationFileContent,
        sourceFileContent,
      );

      return writeFile(destinationFilePath, extendedContent);
    }

    if (isMergeableJsonContent(destinationFileContent)) {
      const extendedContent = extendJson(
        JSON.parse(destinationFileContent),
        JSON.parse(sourceFileContent),
      );

      return writeFile(destinationFilePath, JSON.stringify(extendedContent));
    }

    return writeFile(destinationFilePath, sourceFileContent);
  }
}

main();
