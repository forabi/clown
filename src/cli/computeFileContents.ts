import path from 'path';
import {
  isMergeableJsonContent,
  isDotIgnoreFile,
  readFile,
  getClownConfigPath,
  verifyClownConfigLooksGood,
} from './utils';
import { extendJson } from '../functions/extendJson';
import { extendDotIgnore } from '../functions/extendDotIgnore';
import { getExtensionPathsAndSourceFiles } from './getExtensionPathsAndSourceFiles';
import bluebird from 'bluebird';
import {
  ClownConfig,
  FileContents,
  ExtensionPathAndSourceFiles,
} from './types';
import glob from 'globby';
import fs from 'fs-extra';

export async function computeFileContents(cwd: string) {
  /* for Clown to work, the user has to have a file called `clown.js` at the location
  where they are running the script */
  const clownConfigPath = getClownConfigPath(cwd);
  const clownConfigContent = (await fs.readJson(
    clownConfigPath,
  )) as ClownConfig;

  /* We expect our `clownConfig` to look like this:

  {
    extensions: [
      './node_modules/@hollowverse/config/clown/basics',
      './node_modules/@hollowverse/config/clown/eslint',
      './node_modules/@hollowverse/config/clown/eslintBrowser',
      './node_modules/@hollowverse/config/clown/eslintGraphql'
    ],
  };

  Let's verify that or throw */
  verifyClownConfigLooksGood(clownConfigContent);

  /* So, let's map the list of extension paths so that for each path, we also have a list of the files
  paths relative to the extension path:

  [
    [
      "some/path/node_modules/@hollowverse/config/eslint", [".eslintrc.json", "package.json"],
    ],
    [
      "etc...": ["etc...", "etc..."]
    ]
  ]
  */
  const extensionPathsAndSourceFiles = await bluebird.map(
    clownConfigContent.extensions,
    async (extensionPath: string) => {
      return [
        extensionPath,
        await glob('**', { dot: true, cwd: extensionPath }),
      ] as ExtensionPathAndSourceFiles;
    },
  );

  /* `extensionPathsAndSourceFiles` is a list of all the files inside each extension path.
  Each item in the array looks like this:

  [
    'extensionFolder', // ðŸ‘ˆ this is the extension path
    [
      'package.json', // ðŸ‘ˆ these are the source files
      'eslintrc.json',
      'someFolder/someConfigFile.json'
    ]
  ]

  We want to reduce this structure to a single object where a KEY is the path to the destination file
  and the VALUE is the content of the file as a string.
  */
  const fileContents = await bluebird.reduce(
    extensionPathsAndSourceFiles,
    async (
      fileContents: FileContents,
      [extensionPath, sourceFiles]: ExtensionPathAndSourceFiles,
    ) => {
      const clownConfigPath = getClownConfigPath(cwd);

      await bluebird.each(sourceFiles, async sourceFile => {
        /* We will need to merge each of the source files of this extension path with their targeted
        destination files. That's why we need to loop through the source files. */

        /* First, we need to read the content of the source file, which we will put in the
        destination file */
        const sourceFilePath = path.resolve(extensionPath, sourceFile);
        const sourceFileContent = await readFile(sourceFilePath);

        if (sourceFileContent === undefined) {
          throw new Error(
            `Error loading config source file at ${sourceFilePath}`,
          );
        }

        /* We consider the directory of the Clown config file to be the base from which we resolve
        the location of the destination file. So, if `clown.js` is located at `/some/path/clown.js`
        and the source file is located at `/other/location/package.json` then the location of the
        destination `package.json` is `/some/path/package.json`. That's what the few lines of code
        below compute. */
        const destinationFilePath = path.resolve(
          path.dirname(clownConfigPath),
          sourceFile,
        );

        /* Before we can determine what the content of the destination file will ultimately be,
        we need to read the existing content so that we can merge the content of the source files
        with it. So if we don't yet have a value for the destination file conent, let's assign
        its initial value. */
        if (fileContents[destinationFilePath] === undefined) {
          const destinationFileContent = await readFile(destinationFilePath);

          fileContents[destinationFilePath] = {
            content: destinationFileContent || '',
            type: 'unknown',
          };
        }
        /* Now we have the initial content for that destination file. Let's see if we can merge
        more stuff into it. */

        /* We know how to handle ignore files, such as .gitignore, .npmignore, etc. Let's see if
        our destination file is one of those, and if so, let's merge the content of the source file
        with it. */
        if (isDotIgnoreFile(path.basename(destinationFilePath))) {
          const extendedContent = extendDotIgnore(
            fileContents[destinationFilePath].content,
            sourceFileContent,
          );

          fileContents[destinationFilePath] = {
            content: extendedContent,
            type: 'dotIgnore',
          };
          return;
        }

        /* If the destination file initial content is JSON or the source file content is JSON,
        we treat this file as a JSON and extend it as such. */
        if (
          isMergeableJsonContent(fileContents[destinationFilePath].content) ||
          isMergeableJsonContent(sourceFileContent)
        ) {
          const extendedContent = extendJson(
            JSON.parse(fileContents[destinationFilePath].content || '{}'),
            JSON.parse(sourceFileContent),
          );

          fileContents[destinationFilePath] = {
            content: JSON.stringify(extendedContent),
            type: 'json',
          };
          return;
        }

        /* If the content type of the file is something we don't know how to extend, then we
        just put the content of the source file into the destination as-is */
        fileContents[destinationFilePath] = {
          content: sourceFileContent,
          type: 'unknown',
        };
        return;
      });

      /* When the `each` loop above finishes, that means we have computed all of our file contents.
      Let's return them. */
      return fileContents;
    },
    {},
  );

  /* And from this function we return clownConfig and the fileContents for the clownConfig */
  return {
    clownConfig: { path: clownConfigPath, content: clownConfigContent },
    fileContents,
  };
}
