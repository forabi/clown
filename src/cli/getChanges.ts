import path from 'path';
import {
  iterateOverSourceAndDestinationFiles,
  isMergeableJsonContent,
  writeFile,
  writeJsonFile,
  isDotIgnoreFile,
  readFile,
} from './utils';
import realmkdirp from 'mkdirp';
import { promisify } from 'util';
import { extendJson } from '../functions/extendJson';
import { extendDotIgnore } from '../functions/extendDotIgnore';
import { getExtensionPathsAndSourceFiles } from './getExtensionPathsAndSourceFiles';
import { stripIndents } from 'common-tags';
import bluebird from 'bluebird';
import { Changes, ExtensionPathAndSourceFiles } from './types';

export async function getChanges() {
  const extensionPathsAndSourceFiles = await getExtensionPathsAndSourceFiles();

  return bluebird.reduce(
    extensionPathsAndSourceFiles,
    async (
      changes: Changes,
      [extensionPath, sourceFiles]: ExtensionPathAndSourceFiles,
    ) => {
      await bluebird.each(sourceFiles, async sourceFile => {
        const sourceFilePath = path.resolve(extensionPath, sourceFile);
        const sourceFileContent = await readFile(sourceFilePath);

        if (sourceFileContent === undefined) {
          throw new Error(
            `Error loading config source file at ${sourceFilePath}`,
          );
        }

        const destinationFilePath = path.resolve(process.cwd(), sourceFile);

        if (changes[destinationFilePath] === undefined) {
          const destinationFileContent = await readFile(destinationFilePath);

          changes[destinationFilePath] = {
            content: destinationFileContent || '',
            type: 'unknown',
          };
        }

        if (isDotIgnoreFile(path.basename(destinationFilePath))) {
          const extendedContent = extendDotIgnore(
            changes[destinationFilePath].content,
            sourceFileContent,
          );

          changes[destinationFilePath] = {
            content: extendedContent,
            type: 'dotIgnore',
          };
          return;
        }

        if (
          isMergeableJsonContent(changes[destinationFilePath].content) ||
          isMergeableJsonContent(sourceFileContent)
        ) {
          const extendedContent = extendJson(
            JSON.parse(changes[destinationFilePath].content || '{}'),
            JSON.parse(sourceFileContent),
          );

          changes[destinationFilePath] = {
            content: JSON.stringify(extendedContent),
            type: 'json',
          };
          return;
        }

        changes[destinationFilePath] = {
          content: sourceFileContent,
          type: 'unknown',
        };
        return;
      });

      return changes;
    },
    {},
  );
}
