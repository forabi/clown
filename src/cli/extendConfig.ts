import path from 'path';
import { writeFile, writeJsonFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import fs from 'fs-extra';
import mockFs from 'mock-fs';

export async function extendConfig(
  cwd: string,
  mockContent: { [name: string]: string } | undefined = undefined,
) {
  const { fileContents } = await computeFileContents(cwd);
  const iterableFileContents = _.map(
    fileContents,
    (fileContent, destinationPath) => ({
      destinationPath,
      fileContent,
    }),
  );

  if (mockContent) {
    mockFs(mockContent);
  }

  return await bluebird.each(
    iterableFileContents,
    async ({ destinationPath, fileContent: { content, type } }) => {
      await fs.mkdirp(path.dirname(destinationPath));

      if (type === 'json') {
        return writeJsonFile(destinationPath, JSON.parse(content));
      }

      return writeFile(destinationPath, content);
    },
  );
}
