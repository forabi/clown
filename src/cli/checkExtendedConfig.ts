import { readFile } from './utils';
import { computeFileContents } from './computeFileContents';
import _ from 'lodash';
import bluebird from 'bluebird';
import { patchFs } from 'fs-monkey';
import { vol } from '@forabi/memfs';
import { extendConfig } from './extendConfig';
import { DestinationContent, Discrepancies } from './types';
import { printErrors } from './printErrors';
import fs from 'fs';

export async function checkExtendedConfig(cwd: string) {
  /* We need to check if the content that currently exists on disk is the same as the content
  that Clown thinks should be on disk.

  The first step is to get the content that's currently on disk */
  const {
    fileContents,
    clownConfig: { path: clownConfigPath, content: clownConfigContent },
  } = await computeFileContents(cwd);
  const iterableFileContents = _.map(
    fileContents,
    (fileContent, destinationPath) => ({ destinationPath, fileContent }),
  );
  const currentContent: DestinationContent = await bluebird.reduce(
    iterableFileContents,
    async (
      currentContentAccumulator: DestinationContent,
      { destinationPath },
    ) => {
      currentContentAccumulator[destinationPath] = await readFile(
        destinationPath,
      );

      return currentContentAccumulator;
    },
    {},
  );

  /* Now that we have the content that's currently on disk, we need to find out what Clown thinks
  should be on disk instead. To do that, we'll use `extendConfig` since that's the function that
  knows the answer to our question. But `extendConfig` writes files to the actual disk, and we don't
  want that here. So we will patch the Node filesystem with a mock filesystem, which `extendConfig` will
  write to. */
  const mockDiskContent = {
    [clownConfigPath]: JSON.stringify(clownConfigContent),
    ..._.mapValues(fileContents, fileContent => fileContent.content),
  };
  vol.fromJSON(mockDiskContent);
  (patchFs as any)(vol, fs);
  await extendConfig(cwd, fileContents);

  /* Now, the expected content should have been written to the mock filesystem. Let's read it and
  compare it with the current content that we have retrieved earlier. */
  const discrepancies: Discrepancies = await bluebird.reduce(
    iterableFileContents,
    async (discrepancyAccumulator: Discrepancies, { destinationPath }) => {
      const expectedContent = (await readFile(destinationPath)) as string;
      const destinationContent = currentContent[destinationPath];

      if (
        destinationContent === undefined ||
        destinationContent !== expectedContent
      ) {
        discrepancyAccumulator[destinationPath] = {
          expected: expectedContent,
          received: destinationContent,
        };
      }

      return discrepancyAccumulator;
    },
    {},
  );

  if (_.isEmpty(discrepancies)) {
    process.exit(0);
  } else {
    printErrors(discrepancies);
    process.exit(1);
  }
}
