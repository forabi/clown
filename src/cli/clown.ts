#!/usr/bin/env node
import program from 'commander';

// import { extendConfig } from './extendConfig';
import { checkExtendedConfig } from './checkExtendedConfig';

const cwd = process.cwd();

// function handleExtendConfig() {
//   return extendConfig(cwd);
// }

// program.command('run').action(handleExtendConfig);

program.command('check').action(() => {
  return checkExtendedConfig(cwd);
});

program.parse(process.argv);

// if (program.args.length < 1) {
//   console.log('=\nFILE: clown.ts\nLINE: 22\n=');
//   handleExtendConfig();
// }
