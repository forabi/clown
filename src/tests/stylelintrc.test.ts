import { extendConfig } from '../cli/extendConfig';
import { disk } from '../../mockHelpers/Disk';
import { getJsonContentFromDisk } from '../../mockHelpers/getJsonContentFromDisk';

describe('package.json', () => {
  it('use case 1', async () => {
    const files = {
      '/clown.js': {
        extensions: ['/clownExtensionA'],
      },
      '/clownExtensionA/.stylelintrc': `{
        "rules": {
          "plugin/no-unsupported-browser-features": [true, {
            "ignore": ["css-featurequeries", "outline", "css-appearance", "viewport-units"]
          }]
        }
      }`,
      '/.stylelintrc': `{
        "plugins": [
          "stylelint-no-unsupported-browser-features"
        ],
        "rules": {
          "plugin/no-unsupported-browser-features": [true],
          "at-rule-no-vendor-prefix": [true],
          "media-feature-name-no-vendor-prefix": [true],
          "selector-no-vendor-prefix": [true],
          "property-no-vendor-prefix": [true],
          "value-no-vendor-prefix": [true],
          "property-blacklist": ["margin-bottom", "margin-right", "margin"]
        },
        "ignoreFiles": [
          "./coverage/**/*"
        ]
      }`,
    };

    disk.setContent(files);

    await extendConfig('/');

    expect(await getJsonContentFromDisk('/.stylelintrc')).toMatchSnapshot();
  });
});
