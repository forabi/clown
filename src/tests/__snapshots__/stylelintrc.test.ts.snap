// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`package.json use case 1 1`] = `
"{
  'plugins': [
    'stylelint-no-unsupported-browser-features'
  ],
  'rules': {
    'plugin/no-unsupported-browser-features': [
      true,
      {
        'ignore': [
          'css-featurequeries',
          'outline',
          'css-appearance',
          'viewport-units'
        ]
      }
    ],
    'at-rule-no-vendor-prefix': [
      true
    ],
    'media-feature-name-no-vendor-prefix': [
      true
    ],
    'selector-no-vendor-prefix': [
      true
    ],
    'property-no-vendor-prefix': [
      true
    ],
    'value-no-vendor-prefix': [
      true
    ],
    'property-blacklist': [
      'margin-bottom',
      'margin-right',
      'margin'
    ]
  },
  'ignoreFiles': [
    './coverage/**/*'
  ]
}
"
`;
